### DISCUSSION TITLE
errorHandler not handling global resolver errors/exceptions?

### DISCUSSION THREAD

**Discussion by lucashfreitas** [mercurius-js/mercurius#709] [Q&A] [answered]
Hi There,

Is global error supposed to catch unhandled errors/exception from resolvers? I don't want to handle errors in every resolver itself and trying to handle that and log in a global handler for the graphql operations, but for some reason the `errorHandler` function is not being called whenever I throw an error from the graphql resolvers. 

The code below shows how I am registering the error Handler, therefore I can confirm that is not being called:

``` typescript

 app.register(mercurius, {
    path: options.path,
    routes: true,
    defineMutation: true,
    errorHandler: (
      error: MercuriusError,
      request: FastifyRequest,
      reply: FastifyReply
    ) => {
   // I can confirm this code is not being executed by debugging it and also adding a thumb code like throw new Error("test");
      reply.send({
        data: null,
        errors: [
          {
            message: error.message,
            code: error.code,
          },
        ],
      });
    },

```

Also another question is about what is the recommended way to access the context inside the error handler as I am planning to log custom errors using this function. 

Thanks

**Comment by mcollina** [comment, marked as answer]
You have to handle all errors in your resolvers, it's how you prevent all sort of problems. Note that this concept is true for all Node.js code and not specific to Mercurius.

Could you upload an example to reproduce your problem? So we can help you.

**Comment by lucashfreitas** [comment]
Hey @mcollina ,

I hope you are well! Thank you very much for the quick response.

I actually handle expected errors on all resolvers, therefore for unexpected errors I would like to log it and return a custom generic response e.g, Sorry, something went wrong, please try again.... bla bla bla. So instead of doing that for every resolver I thought of delegating this to a global error handler. 

The main goal is to avoid adding a generic error handler that produces the same output for every single resolver in the code


Here is the resolver code


``` typescript

try {
 //resolver stuff...
}

catch (error) {
 if(error.code === 'knownError') return handleError(err);
 //this error was unexpected.
 throw error;
}

```
Here is the errorHandler code: 

``` typescript

app.register(mercurius, {
    path: options.path,
    routes: true,
    defineMutation: true,
    errorHandler: (
      error: MercuriusError,
      request: FastifyRequest,
      reply: FastifyReply
    ) => {
   //every unhandled/unexpected error would fall into this method.
//I would grab the graphql context/logger service and log the error and return a custom exception
      reply.send({
        data: null,
        errors: [
          {
            message: error.message,
            code: error.code,
          },
        ],
      });
    },


```

**Reply by mcollina** [reply]
I do not undertand. You cannot be safely handle uncaughtException/unhadledRejections without crashing.

**Reply by lucashfreitas** [reply]
@mcollina what happens if an error is not handled in a resolver? How does mercurius handle uncaughtException in order to convert it to a graphql error spec with a json object containing `errors`? Is that handled by native js/graphql library or mercury itself?  

Does mercurius consider that all resolvers will have a try/catch and handle errors?

**Reply by mcollina** [reply]
Mercurius does this already. Could you produce an example where this is not happening?

**Reply by lucashfreitas** [reply]
Yeap, I was wondering if would be possible to override the mercurius handler using `errorHandler` in order to do things such as logging instead of doing that manually on every resolver function? 

I have `pinojs` logger instances in my graphql context, so my idea would be to override the way mercurius handle errors and add a logging step to automatically log unhandledErrors in resolvers.

**Reply by mcollina** [reply]
It is possible.

**Reply by lucashfreitas** [reply]
hmm, that's what I am trying to do as described in the code below, but despite I am trowing an error in my resolver the generic errorHandler is not being called:

``` typescript

app.register(mercurius, {
    path: options.path,
    routes: true,
    defineMutation: true,
    errorHandler: (
      error: MercuriusError,
      request: FastifyRequest,
      reply: FastifyReply
    ) => {
   // I can confirm this code is not being executed by debugging it/console logging, etc.
      reply.send({
        data: null,
        errors: [
          {
            message: error.message,
            code: error.code,
          },
        ],
      });
    },

```

**Reply by mcollina** [reply]
The only way we can help you is for you to produce an example of code that should be working but it's not. All the code you provided are snippet and not something we can run to check - there is either a bug in our code or something in yours. I have asked for such an example twice already. This is the third time and I won't be asking a fourth ([first](https://github.com/mercurius-js/mercurius/discussions/709#discussioncomment-2079459) and [second](https://github.com/mercurius-js/mercurius/discussions/709#discussioncomment-2087978)).

**Reply by lucashfreitas** [reply]
Thanks for your patience. The reason why I kept the discussion is to understand if that's something that is being handled by mercurius - I will provide a repository with a sample. Sorry for taking your time :). Closing it for now.
